dev/src/bulk-writer.ts

```diff
@@ -708,29 +867,44 @@ export class BulkWriter {
    * @private
    */
   private sendReadyBatches(): void {
-    const unsentBatches = this.batchQueue.filter(
-      batch => batch.state === BatchState.READY_TO_SEND
-    );
-
-    let index = 0;
-    while (
-      index < unsentBatches.length &&
-      unsentBatches[index].state === BatchState.READY_TO_SEND
-    ) {
+    // Always send retry batches first if the retryBatchQueue isn't empty.
+    const sendRetries = this._retryBatchQueue.length > 0;
+    const unsentBatches = sendRetries
+      ? this._retryBatchQueue
+      : this._batchQueue;
+    for (let index = 0; index < unsentBatches.length; ++index) {
       const batch = unsentBatches[index];
 
+      // Send all retries whenever possible.
+      if (sendRetries) {
+        batch.markReadyToSend();
+      }
+
+      if (batch.state !== BatchState.READY_TO_SEND) {
+        continue;
+      }
+
+      const batchDeferred = new Deferred<void>();
+      this._pendingBatches.add(batchDeferred.promise);
+
       // Send the batch if it is under the rate limit, or schedule another
       // attempt after the appropriate timeout.
       const delayMs = this.rateLimiter.getNextRequestDelayMs(batch.opCount);
       assert(delayMs !== -1, 'Batch size should be under capacity');
       if (delayMs === 0) {
-        this.sendBatch(batch);
+        this.sendBatch(batch, sendRetries)
+          .then(batchDeferred.resolve, batchDeferred.reject)
+          .then(() => {
+            this._pendingBatches.delete(batchDeferred.promise);
+          });
       } else {
-        delayExecution(() => this.sendReadyBatches(), delayMs);
+        delayExecution(() => {
+          batchDeferred.resolve();
+          this._pendingBatches.delete(batchDeferred.promise);
+          this.sendReadyBatches();
```

Comment:
Please add a comment here that explain that `batchDeferred` in this case isn't actually a batch, but rather just an operation that waits for the timeout to fire, which will ad more batches.